datafusion/src/execution/runtime_env.rs

```diff
@@ -25,47 +25,110 @@ use crate::{
+    PARQUET_PRUNING, REPARTITION_AGGREGATIONS, REPARTITION_JOINS, REPARTITION_WINDOWS,
+    TARGET_PARTITIONS,
+};
+use crate::prelude::{SessionConfig, SessionContext};
+use datafusion_common::DataFusionError;
+use once_cell::sync::OnceCell;
+use parking_lot::Mutex;
+use std::path::PathBuf;
 use std::sync::Arc;
 
-#[derive(Clone)]
-/// Execution runtime environment. This structure is passed to the
-/// physical plans when they are run.
+/// Global singleton RuntimeEnv
+pub static RUNTIME_ENV: OnceCell<Arc<RuntimeEnv>> = OnceCell::new();
+
+/// Execution runtime environment. This structure is a singleton for each Scheduler/Executor instance.
 pub struct RuntimeEnv {
-    /// Default batch size while creating new batches
-    pub batch_size: usize,
+    /// Executor Id
+    pub executor_id: Option<String>,
+    /// Local Env
+    pub is_local: bool,
     /// Runtime memory management
     pub memory_manager: Arc<MemoryManager>,
     /// Manage temporary files during query execution
     pub disk_manager: Arc<DiskManager>,
+    /// Object Store that are registered within the Scheduler's or Executors' Runtime
+    pub object_store_registry: Arc<ObjectStoreRegistry>,
+    /// DataFusion task contexts that are registered within the Executors' Runtime
+    pub task_context_registry: Option<Arc<TaskContextRegistry>>,
+    /// DataFusion session contexts that are registered within the Scheduler's Runtime
+    pub session_context_registry: Option<Arc<SessionContextRegistry>>,
 }
 
-impl Debug for RuntimeEnv {
-    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
-        write!(f, "RuntimeEnv")
+impl RuntimeEnv {
+    /// Create an executor env based on configuration
+    pub fn new_executor_env(config: RuntimeConfig, executor_id: String) -> Result<Self> {
+        let RuntimeConfig {
+            memory_manager,
+            disk_manager,
+        } = config;
+        Ok(Self {
+            executor_id: Some(executor_id),
+            is_local: false,
+            memory_manager: MemoryManager::new(memory_manager),
+            disk_manager: DiskManager::try_new(disk_manager)?,
+            object_store_registry: Arc::new(ObjectStoreRegistry::new()),
+            task_context_registry: Some(Arc::new(TaskContextRegistry::new())),
+            session_context_registry: None,
+        })
     }
-}
 
-impl RuntimeEnv {
-    /// Create env based on configuration
-    pub fn new(config: RuntimeConfig) -> Result<Self> {
+    /// Create a scheduler env based on configuration
+    pub fn new_scheduler_env(config: RuntimeConfig) -> Result<Self> {
         let RuntimeConfig {
-            batch_size,
             memory_manager,
             disk_manager,
         } = config;
+        Ok(Self {
+            executor_id: None,
+            is_local: false,
+            memory_manager: MemoryManager::new(memory_manager),
+            disk_manager: DiskManager::try_new(disk_manager)?,
+            object_store_registry: Arc::new(ObjectStoreRegistry::new()),
```

Comment:
I like the idea of adding a `object_store_registry`, `task_context_registry` and `session_conctext_registry` to the `RuntimeEnvironment` -- I think having a way to track those items is a good one. 👍